Spark 4.0 integration

huaxingao opened the pull request.

huaxingao commented on build.gradle:
```diff
@@ -119,6 +119,9 @@ allprojects {
     mavenCentral()
     mavenLocal()
+    maven {
+      url "https://repository.apache.org/content/repositories/snapshots/"
```
Need this for now because this PR uses `4.0.1-SNAPSHOT`. 
Will remove this in the future

huaxingao commented on gradle/libs.versions.toml:
```diff
@@ -81,6 +82,7 @@ slf4j = "2.0.17"
 snowflake-jdbc = "3.23.2"
 spark34 = "3.4.4"
 spark35 = "3.5.5"
+spark40 = "4.0.1-SNAPSHOT"
```
Will change to official released Spark 4.0 in the future.

RussellSpitzer replied:
Shouldn't this be the RC now?

huaxingao replied:
I will change this to RC4.

huaxingao commented on gradle/libs.versions.toml:
```diff
@@ -24,6 +24,7 @@ activation = "1.1.1"
 aliyun-sdk-oss = "3.10.2"
 analyticsaccelerator = "1.0.0"
 antlr = "4.9.3"
+antlr413 = "4.13.1" # For Spark 4.0 support
```
to be consistent with antlr version in Spark 4.0

huaxingao commented on build.gradle:
```diff
@@ -760,6 +763,7 @@ project(':iceberg-hive-metastore') {
     testImplementation project(path: ':iceberg-api', configuration: 'testArtifacts')
     testImplementation project(path: ':iceberg-core', configuration: 'testArtifacts')
     testImplementation libs.awaitility
+    testImplementation 'com.zaxxer:HikariCP:5.1.0'
```
Spark 4.0 removed dependency on `com.jolbox:bonecp`. In `TestHiveMetastore.java`, I explicitly set the connection pool to `conf.set("datanucleus.connectionPoolingType", "HikariCP");`, otherwise, it will default to ``com.jolbox:bonecp` and we will get 
```
Exception in thread "TThreadPoolServer WorkerProcess-5" java.lang.NoClassDefFoundError: com/jolbox/bonecp/BoneCPConfig
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.setupJdbcConnectionPool(TxnHandler.java:3156)
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.setConf(TxnHandler.java:260)
```
Since I explicitly set the connection pool to `conf.set("datanucleus.connectionPoolingType", "HikariCP");`, we need to add the test dependency.

pan3793 replied:
DBCP(already pulled by Spark, see SPARK-48538) can be used to avoid pulling HikariCP dependency

huaxingao replied:
somehow after I changed to `DBCP`, I started to have 
```
> Task :iceberg-spark:iceberg-spark-runtime-4.0_2.13:integrationTest FAILED
```
It doesn't have any additional information. 

If I run  `./gradlew :iceberg-spark:iceberg-spark-runtime-4.0_2.13:integrationTest` on my local, it always passes.

Now even after I switched back to `HikariCP`. `Task :iceberg-spark:iceberg-spark-runtime-4.0_2.13:integrationTest ` still failed.

pan3793 replied:
should be irrelevant, could you try uploading the reports for the failure test?

`.github/workflows/spark-ci.yml`
```patch
      - uses: actions/upload-artifact@v4
        if: failure()
        with:
          name: test logs
          path: |
            **/build/testlogs
+           **/build/reports
```

pan3793 replied:
what's the root cause?

huaxingao replied:
```
Caused by: org.junit.platform.commons.JUnitException: OutputDirectoryProvider not available; probably due to unaligned versions of the junit-platform-engine and junit-platform-launcher jars on the classpath/module path.
```
I guess `junit-platform-engine` has been upgraded in upstream. I rebased and the problem went away.

huaxingao commented on core/src/main/java/org/apache/iceberg/MetadataColumns.java:
```diff
@@ -50,7 +50,7 @@ private MetadataColumns() {}
   public static final String SPEC_ID_COLUMN_DOC = "Spec ID used to track the file containing a row";
   public static final NestedField SPEC_ID =
-      NestedField.required(
+      NestedField.optional(
```
In Spark 4.0, the metadata columns are nullable, so need to change this field to optional. 
https://github.com/apache/spark/pull/50246#discussion_r2023825620

huaxingao commented on hive-metastore/src/test/java/org/apache/iceberg/hive/TestHiveMetastore.java:
```diff
@@ -282,6 +282,8 @@ private void initConf(HiveConf conf, int port, boolean directSql) {
     conf.set(
         HiveConf.ConfVars.HIVE_IN_TEST.varname, HiveConf.ConfVars.HIVE_IN_TEST.getDefaultValue());
+
+    conf.set("datanucleus.connectionPoolingType", "HikariCP");
```
Spark 4.0 removed dependency on com.jolbox:bonecp.  If not set the cp type, it will default to `com.jolbox:bonecp` and we will get
```
Exception in thread "TThreadPoolServer WorkerProcess-5" java.lang.NoClassDefFoundError: com/jolbox/bonecp/BoneCPConfig
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.setupJdbcConnectionPool(TxnHandler.java:3156)
	at org.apache.hadoop.hive.metastore.txn.TxnHandler.setConf(TxnHandler.java:260)
```

huaxingao commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckViews.scala:
```diff
@@ -105,7 +106,7 @@ object CheckViews extends (LogicalPlan => Unit) {
     val newCyclePath = cyclePath :+ currentViewIdent
     if (currentViewIdent == viewIdent) {
-      throw new AnalysisException(String.format("Recursive cycle in view detected: %s (cycle: %s)",
+      throw new IcebergAnalysisException(String.format("Recursive cycle in view detected: %s (cycle: %s)",
```
Spark 4.0 doesn't allow constructing an AnalysisException with error message only. We have to provide an error class. In order to get around this, I have added `IcebergAnalysisException`, which only needs an error message.

huaxingao commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/IcebergCall.scala:
```diff
@@ -0,0 +1,34 @@
+import org.apache.spark.sql.catalyst.util.truncatedString
+import org.apache.spark.sql.connector.iceberg.catalog.Procedure
+
+case class IcebergCall(procedure: Procedure, args: Seq[Expression]) extends LeafCommand {
```
Spark 4.0 has its own implementation of `Call`. In this Spark 4.0/Iceberg integration PR, I simply change `Call` to I`cebergCall` to avoid name collision. In the future, will need to migrate to Spark's native Call framework.

huaxingao commented on spark/v4.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/ExtensionsTestBase.java:
```diff
@@ -43,7 +43,7 @@ public static void startMetastoreAndSpark() {
     TestBase.hiveConf = metastore.hiveConf();
 
-    TestBase.spark.close();
+    TestBase.spark.stop();
```
don't have this `sql.SparkSession.close()` any more so replace with `stop`

huaxingao commented on spark/v4.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestCallStatementParser.java:
```diff
@@ -69,16 +70,9 @@ public static void stopSpark() {
-            });
+  public void testDelegateUnsupportedProcedure() throws ParseException {
+    LogicalPlan plan = parser.parsePlan("CALL cat.d.t()");
+    assertThat(plan.toString().contains("CALL cat.d.t()"));
```
Since Spark 4.0 has Stored Procedure support, so this doesn't throw ParseException any more

RussellSpitzer replied:
Shouldn't we just remove this test then? I'm not sure what we are checking here any more?

huaxingao commented on spark/v4.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestSystemFunctionPushDownInRowLevelOperations.java:
```diff
@@ -260,7 +267,12 @@ private void checkUpdate(RowLevelOperationMode mode, String cond) {
+          try {
+            changeDF.coalesce(1).writeTo(tableName(CHANGES_TABLE_NAME)).create();
+          } catch (TableAlreadyExistsException e) {
+            throw new AlreadyExistsException(
```
```
  @throws(classOf[TableAlreadyExistsException])
  def create(): Unit
```
This API now throws `TableAlreadyExistsException`

RussellSpitzer replied:
Why do we need to change the exception type?

huaxingao replied:
if I re-throw without doing anything
```
          try {
            changeDF.coalesce(1).writeTo(tableName(CHANGES_TABLE_NAME)).create();
          } catch (TableAlreadyExistsException e) {
            throw e;
          }
```
I got `Unhandled exception: org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException`

RussellSpitzer replied:
Ah got it!

huaxingao commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/procedures/CreateChangelogViewProcedure.java:
```diff
@@ -246,11 +249,10 @@ private String viewName(ProcedureInput input, String tableName) {
     Column[] sortSpec = sortSpec(df, repartitionSpec, false);
     StructType schema = df.schema();
-    String[] identifierFields =
-        Arrays.stream(repartitionSpec).map(Column::toString).toArray(String[]::new);
```
`Column::toString` now returns a fully qualified column, and it causes `fieldIndex(name: String)` to fail later. 

We actually have a String array of `identifierFields` in `computeUpdateImages` (only need to append `MetadataColumns.CHANGE_ORDINAL.name()` in the end). So Instead of calling 
```
override def toString: String = node.sql
```
I think it's more efficient to pass in the String array of `identifierFields`

huaxingao commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:
```diff
@@ -875,7 +875,13 @@ private static void deleteManifests(FileIO io, List<ManifestFile> manifests) {
     SparkTable sparkTable = new SparkTable(table, snapshotId, false);
     DataSourceV2Relation relation = createRelation(sparkTable, ImmutableMap.of());
-    return Dataset.ofRows(spark, relation);
+    if (!(spark instanceof org.apache.spark.sql.classic.SparkSession)) {
```
Spark 4.0 has [abstract class SparkSession](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala#L63) and [classic.SparkSession](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala#L63). We will always try to use the general one, and only cast to `classic.SparkSession` when the private APIs are needed.

The same rule applies to `Dataset`

RussellSpitzer replied:
I think I noted other places, but let's change all of these checks to a Precondition if we can

huaxingao commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkUtil.java:
```diff
@@ -154,8 +154,9 @@ public static Configuration hadoopConfCatalogOverrides(SparkSession spark, Strin
     final Configuration conf = spark.sessionState().newHadoopConf();
+
     spark
-        .sqlContext()
```
The generic [abstract class SparkSession](https://github.com/apache/spark/blob/master/sql/api/src/main/scala/org/apache/spark/sql/SparkSession.scala#L63) doesn't have this API `sqlContext` any more

huaxingao commented:
cc @RussellSpitzer @szehon-ho @amogh-jahagirdar @nastra 
Could you please take a look at this PR when you have a moment? Thanks a lot!

RussellSpitzer commented on .github/workflows/java-ci.yml:
```diff
@@ -95,7 +95,7 @@ jobs:
     runs-on: ubuntu-22.04
     strategy:
       matrix:
-        jvm: [11, 17, 21]
```
This turns 11 off for our earlier versions, are we able to check with 11 just skipped for 4.0? I don't think its a huge loss if we stop testing 11 but it would be better if we could maintain that

pan3793 replied:
Java 11 compatibility is still guarded by `javac`/`scalac` compile options `-release:11`, in other words, jar compiled via Java 17/21 still works with Java 11 runtime.
https://github.com/apache/iceberg/blob/013d09e473faada1a3b21c155964ec4d1254d69d/build.gradle#L189-L192
https://github.com/apache/iceberg/blob/013d09e473faada1a3b21c155964ec4d1254d69d/build.gradle#L238-L246

huaxingao replied:
I'll look into this and see if we can still run checks with Java 11, just skipping it specifically for 4.0.

RussellSpitzer commented on settings.gradle:
```diff
@@ -161,6 +161,18 @@ if (sparkVersions.contains("3.5")) {
 }
 
+if (sparkVersions.contains("4.0")) {
+  include ":iceberg-spark:spark-4.0_2.13"
```
Does this theoretically create Spark 4.0 if Scala 2.12 is set without error? Feels like we should still use scalaVersion here but have an assert > 2.12 or something.

huaxingao replied:
The problem is that Flink does work with scala 2.13. I use 2.13 to build Spark module only. I think we can't assert > 2.12 because we still need 2.12 to build Flink.

RussellSpitzer replied:
Doesn't that mean when Scala is set to 2.12 we just should allow building Spark 4.0?

RussellSpitzer commented on gradle/libs.versions.toml:
```diff
@@ -92,6 +94,8 @@ aliyun-sdk-oss = { module = "com.aliyun.oss:aliyun-sdk-oss", version.ref = "aliy
 antlr-antlr4 = { module = "org.antlr:antlr4", version.ref = "antlr" }
 antlr-runtime = { module = "org.antlr:antlr4-runtime", version.ref = "antlr" }
+antlr-antlr413 = { module = "org.antlr:antlr4", version.ref = "antlr413" }
+antlr-runtime413 = { module = "org.antlr:antlr4-runtime", version.ref = "antlr413" }
```
This feels a little odd to me, Do we really need a second antlr version? Is there incompatibility with the runtime?

pan3793 replied:
better to align with Spark, otherwise at least a warning message printed from the ANTLR-generated code at runtime.

huaxingao replied:
We still need the original `antlr = "4.9.3"` to build Spark 3.4 and 3.5.

RussellSpitzer commented on spark/v4.0/build.gradle:
```diff
@@ -181,14 +182,15 @@ project(":iceberg-spark:iceberg-spark-extensions-${sparkMajorVersion}_${scalaVer
     testImplementation libs.parquet.hadoop
     testImplementation libs.awaitility
-    testImplementation "org.apache.datafusion:comet-spark-spark${sparkMajorVersion}_${scalaVersion}:0.5.0"
+    testImplementation "org.apache.datafusion:comet-spark-spark3.5_2.13:0.5.0"
```
Why are we using the old version of comet here? Just because the new one isn't available?

huaxingao replied:
Rignt, we don't have a Spark 4.0 Comet yet.

RussellSpitzer commented on spark/v4.0/spark-extensions/src/test/java/org/apache/iceberg/spark/extensions/TestViews.java:
```diff
@@ -538,7 +538,7 @@ public void readFromViewReferencingTempFunction() throws NoSuchTableException {
     assertThatThrownBy(() -> sql("SELECT * FROM %s", viewName))
         .isInstanceOf(AnalysisException.class)
-        .hasMessageStartingWith(expectedErrorMsg);
+        .hasMessageContaining((expectedErrorMsg));
```
Extra parens

huaxingao replied:
Will remove

RussellSpitzer commented on spark/v4.0/spark/src/jmh/java/org/apache/iceberg/spark/action/DeleteOrphanFilesBenchmark.java:
```diff
@@ -59,8 +59,8 @@
- * <p>To run this benchmark for spark-3.5: <code>
- *   ./gradlew -DsparkVersions=3.5 :iceberg-spark:iceberg-spark-3.5_2.12:jmh
+ * <p>To run this benchmark for spark-4.0: <code>
+ *   ./gradlew -DsparkVersions=4.0 :iceberg-spark:iceberg-spark-4.0_2.12:jmh
```
scala version should be 2.13

huaxingao replied:
Will change. Thanks for catching this problem!

RussellSpitzer commented on spark/v4.0/spark/src/jmh/java/org/apache/iceberg/spark/data/parquet/SparkParquetReadersFlatDataBenchmark.java:
```diff
@@ -58,8 +58,8 @@
- * <p>To run this benchmark for spark-3.5: <code>
- *   ./gradlew -DsparkVersions=3.5 :iceberg-spark:iceberg-spark-3.5_2.12:jmh
+ * <p>To run this benchmark for spark-4.0: <code>
+ *   ./gradlew -DsparkVersions=4.0 :iceberg-spark:iceberg-spark-4.0_2.12:jmh
```
Scala version

RussellSpitzer commented on spark/v4.0/spark/src/jmh/java/org/apache/iceberg/spark/source/parquet/vectorized/VectorizedReadParquetDecimalBenchmark.java:
```diff
@@ -48,8 +48,8 @@
- * <p>To run this benchmark for spark-3.5: <code>
- *   ./gradlew -DsparkVersions=3.5 :iceberg-spark:iceberg-spark-3.5_2.12:jmh \
+ * <p>To run this benchmark for spark-4.0: <code>
+ *   ./gradlew -DsparkVersions=4.0 :iceberg-spark:iceberg-spark-4.0_2.12:jmh \
```
Scala version wrong in all above docs as well

RussellSpitzer commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkConfParser.java:
```diff
@@ -40,7 +40,7 @@ class SparkConfParser {
   SparkConfParser() {
     this.properties = ImmutableMap.of();
-    this.sessionConf = new RuntimeConfig(SQLConf.get());
+    this.sessionConf = new org.apache.spark.sql.classic.RuntimeConfig(SQLConf.get());
```
Why do we have both sql.RuntimeConfig and sql.classic.RuntimeConfig ?

huaxingao replied:
I tried to use `sql.RuntimeConfig` as much as possible and only switch to `sql.classic.RuntimeConfig` when necessary. That's why I have both `sql.RuntimeConfig` and `sql.classic.RuntimeConfig`.

RussellSpitzer commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/SparkTableUtil.java:
```diff
@@ -875,7 +875,13 @@ private static void deleteManifests(FileIO io, List<ManifestFile> manifests) {
-    return Dataset.ofRows(spark, relation);
+    if (!(spark instanceof org.apache.spark.sql.classic.SparkSession)) {
+      throw new IllegalArgumentException(
+          "spark is supposed to be org.apache.spark.sql.classic.SparkSession");
```
Wouldn't this be something like "Cannot use loadTable with Spark Connect" ?

huaxingao replied:
We probably don't need to take consideration of Spark Connect? I guess we don't need to run Iceberg on client side?

RussellSpitzer commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteDataFilesSparkAction.java:
```diff
@@ -106,7 +106,7 @@ public class RewriteDataFilesSparkAction
 
   RewriteDataFilesSparkAction(SparkSession spark, Table table) {
-    super(spark.cloneSession());
+    super(((org.apache.spark.sql.classic.SparkSession) spark).cloneSession());
```
I'm not sure this is safe to do, again I think we may have an issue with Spark Connect right? Don't we need to protect against these casts?

pan3793 replied:
the cast should be safe, in spark connect arch, the current iceberg spark connector is a side server plugin, it is not supposed to be invoked on the client side.

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestComputeTableStatsAction.java:
```diff
@@ -392,7 +392,13 @@ private Dataset<Row> randomDataDF(Schema schema) {
-    return spark.internalCreateDataFrame(JavaRDD.toRDD(rowRDD), rowSparkType, false);
+    if (!(spark instanceof org.apache.spark.sql.classic.SparkSession)) {
+      throw new IllegalArgumentException(
+          "spark is supposed to be org.apache.spark.sql.classic.SparkSession");
```
Error message again

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteDataFilesAction.java:
```diff
@@ -1698,6 +1698,7 @@ public void testZOrderSort() {
 
   @TestTemplate
   public void testZOrderAllTypesSort() {
+    spark.conf().set("spark.sql.ansi.enabled", "false");
```
Just wondering what we are doing here that isn't ansi compatible

RussellSpitzer commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/source/TestPositionDeletesTable.java:
```diff
@@ -1191,27 +1191,28 @@ public void testWritePartitionEvolutionDisallowed() throws Exception {
               .option(SparkReadOptions.FILE_OPEN_COST, Integer.MAX_VALUE)
               .load(posDeletesTableName);
       assertThat(scanDF.javaRDD().getNumPartitions()).isEqualTo(1);
+      assertThat(!Arrays.asList(scanDF.columns()).contains("partition"));
```
Why do we need this assert?

huaxingao replied:
I can't remember why I added this line. I will remove this.

RussellSpitzer commented:
Took a first pass, I think our main issue here is going to be differentiating between methods/functionality that work in Spark Connect and that which does not. We should make it clear and have some follow up items to try to get some of our *classic* code updated to *connect* when possible. I think almost anywhere we aren't touching the RDD's there should be an upgrade path.

huaxingao commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/spark/sql/catalyst/analysis/ProcedureArgumentCoercion.scala:
```diff
@@ -19,15 +19,14 @@
 object ProcedureArgumentCoercion extends Rule[LogicalPlan] {
   override def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators {
-    case c @ Call(procedure, args) if c.resolved =>
+    case c @ IcebergCall(procedure, args) if c.resolved =>
```
Spark 4.0 has its own [`org.apache.spark.sql.catalyst.plans.logical.Call`](https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/v2Commands.scala#L1600). I have to rename this to `IcebergCall`to avoid name collision. This is only a temporary solution. I will have follow up PR to change the iceberg store procedures to follow Spark 4.0 framework stored procedure framework.

huaxingao commented on spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/sql/TestAggregatePushDown.java:
```diff
@@ -54,8 +54,6 @@ public static void startMetastoreAndSpark() {
     metastore.start();
     TestBase.hiveConf = metastore.hiveConf();
 
-    TestBase.spark.close();
```
Will change to `TestBase.spark.stop()`

huaxingao commented:
Thanks @RussellSpitzer for your review! I’ve addressed your comments where possible, but haven’t had time to update the code yet. After this PR is done, I’ll have followup PRs to eliminate the use of `classic` as much as possible.

huaxingao commented:
@RussellSpitzer @pan3793 
I have addressed the comments. Could you please take one more look when you have time?

pan3793 commented:
The git commits need to be tidied up before merging. Or, commit
1. mv spark-3.5 to spark-4.0
2. copy spark-4.0 spark-3.5

to main first, the spark-4.0 folder won't be recognized by gradle, so it won't fail CI, then focus on this PR and don't need to worry about the git history

huaxingao commented:
> The git commits need to be tidied up before merging. Or, commit
> 
> mv spark-3.5 to spark-4.0
> copy spark-4.0 spark-3.5

Right, when this PR is ready, I will squash the remaining commits so that, in the end, there will be only three commits:
[Move 3.5 as 4.0]
[Copy back 4.0 as 3.5]
[initial support for Spark 4.0]

pan3793 reviewed (approved):
I think it's already in good shape, I suggest merging it first and polishing in follow-up PRs.

pan3793 commented:
BTW, I have compiled your yesterday's version and deployed it with Spark 4.0.0 RC4 in a small YARN cluster, run some TPC-H queries, and it works as expected.

amogh-jahagirdar commented:
Just a fyi, I've locally built off of these changes to prototype spark 4.0 + row lineage and things are looking good! I got the tests I added in https://github.com/apache/iceberg/pull/12736 passing

szehon-ho commented on spark/v4.0/spark/src/main/java/org/apache/iceberg/spark/procedures/CreateChangelogViewProcedure.java:
```diff
@@ -0,0 +1,321 @@
+
+    Column[] repartitionSpec = new Column[identifierColumns.length + 1];
+    for (int i = 0; i < identifierColumns.length; i++) {
+      repartitionSpec[i] = df.col(CreateChangelogViewProcedure.delimitedName(identifierColumns[i]));
```
just wondering, can you explain a bit this diff from 3.5?

szehon-ho replied:
never mind, i see it:  https://github.com/apache/iceberg/pull/12494#discussion_r2040874901

szehon-ho reviewed (approved):
I diffed 3.5 vs 4.0, the changes are quite minimal, thanks a lot @huaxingao !

aihuaxu commented:
@huaxingao  I diff 3.5 and 4.0 and see some changes and seems they are from the recent changes that we need to bring in. Otherwise looks good to me.

RussellSpitzer reviewed (approved):
I think we should get this in even before Spark 4.0 official release because I see we are having some trouble with keeping things in sync now. I think this is very close and most other things can be follow ups.

aihuaxu commented:
> I think we should get this in even before Spark 4.0 official release because I see we are having some trouble with keeping things in sync now. I think this is very close and most other things can be follow ups.

Agree. Seems we already have some challenges to keep sync with 3.5.

amogh-jahagirdar reviewed (approved):
I agree with @RussellSpitzer that it's better to get this in even before the Spark 4.0 release and iterate on it since it's naturally getting difficult to keep things in sync. There aren't any fundamental issues based on my review, mostly just small follow ups. @huaxingao there's a conflcit in gradle.properties, once that's fixed, I'll go ahead and merge!

amogh-jahagirdar commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala:
```diff
@@ -0,0 +1,55 @@
+    extensions.injectResolutionRule { spark => ResolveViews(spark) }
+    extensions.injectResolutionRule { _ => ProcedureArgumentCoercion }
+    extensions.injectCheckRule(_ => CheckViews)
+    extensions.injectResolutionRule { _ => RewriteUpdateTableForRowLineage}
+    extensions.injectResolutionRule { _ => RewriteMergeIntoTableForRowLineage}
```
Discussed offline with @huaxingao , since row lineage in 4.0 will have a very different implementation, we will remove the changes from https://github.com/apache/iceberg/pull/12736/files that are included in this changes.

amogh-jahagirdar commented on spark/v4.0/spark-extensions/src/main/scala/org/apache/iceberg/spark/extensions/IcebergSparkSessionExtensions.scala:
```diff
@@ -0,0 +1,55 @@
+    extensions.injectResolutionRule { spark => ResolveViews(spark) }
+    extensions.injectResolutionRule { _ => ProcedureArgumentCoercion }
+    extensions.injectCheckRule(_ => CheckViews)
+    extensions.injectResolutionRule { _ => RewriteUpdateTableForRowLineage}
```
Discussed offline with @huaxingao , since row lineage in 4.0 will have a very different implementation, we will remove the changes from https://github.com/apache/iceberg/pull/12736/files that are included in this changes.

huaxingao commented:
@amogh-jahagirdar I have removed the changes from https://github.com/apache/iceberg/pull/12736/files for Spark 4.0. I have checked all the recent commits after my yesterday's rebase, the only one that has Spark changes are my Comet bump version [PR](https://github.com/apache/iceberg/commit/94ba295960a5da1ce87c35a1e28353f4a71ccd72). I will manually port the changes into this PR. After the all the checks pass, we can merge.

amogh-jahagirdar commented:
Ok, thank you @huaxingao for your dilligence on keeping up with this PR, I'll go ahead and merge and we can iterate on any follow ups. When the official spark 4.0 release is performed, we would do the dependency changes then. Thank you @RussellSpitzer @pan3793 @szehon-ho for reviewing.

huaxingao commented:
Thank you all! @amogh-jahagirdar @RussellSpitzer @pan3793 @szehon-ho @aihuaxu

pan3793 commented:
Congrats on merging to trunk, but alright, the commit history is lost eventually

huaxingao commented:
Sorry I forgot to squash the commits. I will fix this

huaxingao commented:
I will revert the PR for now
